src/main/resources/com/tresys/nitf/xsd/nitf.dfdl.xsd

```diff
@@ -853,50 +833,72 @@ The message root is 'NITF'.
-      <xs:element name="DeclassificationDate" type="nitf:ECS-A" dfdl:length="8" nillable="true" dfdl:ref="nitf:nilString" />
-      <xs:element name="DeclassificationExemption" type="nitf:ECS-A" dfdl:length="4" nillable="true" dfdl:ref="nitf:nilString" />
-      <xs:element name="Downgrade" type="nitf:ECS-A" dfdl:length="1" nillable="true" dfdl:ref="nitf:nilString" />
-      <xs:element name="DowngradeDate" type="nitf:date" nillable="true" dfdl:ref="nitf:nilDate" />
-      <xs:element name="ClassificationText" type="nitf:ECS-A" dfdl:length="43" nillable="true" dfdl:ref="nitf:nilString" />
-      <xs:element name="ClassificationAuthorityType" type="nitf:ECS-A" dfdl:length="1" nillable="true" dfdl:ref="nitf:nilString" />
-      <xs:element name="ClassificationAuthority" type="nitf:ECS-A" dfdl:length="40" nillable="true" dfdl:ref="nitf:nilString" />
-      <xs:element name="ClassificationReason" type="nitf:ECS-A" dfdl:length="1" nillable="true" dfdl:ref="nitf:nilString" />
-      <xs:element name="SecuritySourceDate" type="nitf:date" nillable="true" dfdl:ref="nitf:nilDate" />
-      <xs:element name="SecurityControlNumber" type="nitf:ECS-A" dfdl:length="15" nillable="true" dfdl:ref="nitf:nilString" />
-    </xs:sequence>
+      <xs:element name="ClassificationInfo2.0">
+        <xs:complexType>
+          <xs:sequence>
+            <xs:element name="SecurityClassification" dfdl:length="1">
+              <xs:simpleType>
+                <xs:restriction base="nitf:ECS-A">
+                  <xs:enumeration value="T"/>
+                  <xs:enumeration value="S"/>
+                  <xs:enumeration value="C"/>
+                  <xs:enumeration value="R"/>
+                  <xs:enumeration value="U"/>
+                </xs:restriction>
+              </xs:simpleType>
+            </xs:element>
+            <xs:element name="Codewords" type="nitf:BCS-A" dfdl:length="40" nillable="true" dfdl:ref="nitf:nilString" />
+            <xs:element name="ControlAndHandling" type="nitf:ECS-A" dfdl:length="40" nillable="true" dfdl:ref="nitf:nilString" />
+            <xs:element name="ReleasingInstructions" type="nitf:ECS-A" dfdl:length="40" nillable="true" dfdl:ref="nitf:nilString" />
+            <xs:element name="ClassificationAuthority" type="nitf:ECS-A" dfdl:length="20" nillable="true" dfdl:ref="nitf:nilString" />
+            <xs:element name="SecurityControlNumber" type="nitf:ECS-A" dfdl:length="20" nillable="true" dfdl:ref="nitf:nilString" />
+            <xs:element name="Downgrade" type="nitf:ECS-A" dfdl:length="6" nillable="true" dfdl:ref="nitf:nilString" />
+            <xs:element name="DowngradeEvent" type="nitf:ECS-A" dfdl:length="40" nillable="true" dfdl:ref="nitf:nilString"
+            minOccurs="0" dfdl:occursCountKind="expression" dfdl:occursCount="{ if (fn:nilled(../Downgrade)) then 0 else if (../Downgrade eq '999998') then 1 else 0 }" />
+          </xs:sequence>
+        </xs:complexType>
+      </xs:element>
+    </xs:choice>
```

Comment:
Similar to the header info, would it be possible to avoid the ClassificationInfoset2.0/2.1 elements, and just have some elements be optional or runtime compute lengths? Or maybe the choice can just be a choice of sequence's instead of a choice of elements? That might be easier, but I think you'd need to move SecurityClassification outside of the choice to avoid UPA violation.

Reply:
This was one of the things I was hoping to discuss during the code review.  I knew this 2.0 vs 2.1 stuff was kind of ugly, but I wasn't sure what the best way to cleanly handle it would be.  I'll look into the choice of sequences first.